### PR TITLE
Count Intersection task: enable model upload to Hugging Face Hub

### PR DESCRIPTION
**Pull request by oliveirabruno01**
This PR adds the ability to upload a fine-tuned model for the `count_intersection` task to the Hugging Face Hub, as discussed at #5.  

Changes currently include:  
- Updated the `count_intersection` task to include model upload functionality.
- Refactored the task's notebook for better readability.
- Made minor changes to streamline the workflow.

### CODE REVIEW

**Comment by oliveirabruno01** [comment]
@ariG23498

File: count_intersection/create_intersection_dataset.py [hunk 1]
```diff
@@ -110,7 +114,7 @@ def main(
     # push to hub
     if push_to_hub:
         ds = load_dataset(dataset_folder)
-        ds.push_to_hub("ariG23498/intersection-dataset")
+        ds.push_to_hub(dataset_id)
```
Comment (ariG23498) [inline count_intersection/create_intersection_dataset.py:118]: Neat!

File: count_intersection/create_intersection_dataset.py
Comment (ariG23498) [inline count_intersection/create_intersection_dataset.py:1]: Can we also update the `Example usage` section right at the beginning of the script?
That would help people copy the command and run it. (Not a blocker, but good to have)

File: count_intersection/README.md [hunk 2]
```diff
@@ -37,22 +39,32 @@ python create_intersection_dataset.py \
   - `dataset_folder`: Directory to save the dataset.
   - `dataset_split`: Dataset split (`train` or `validation`).
   - `push_to_hub`: Whether to upload the dataset to the Hugging Face Hub.
+  - `dataset_id`: ID (username + dataset name) used to upload to the Hub. Required if `push_to_hub` is True.
```
Comment (ariG23498) [inline count_intersection/README.md]: Would be a nice idea to show an example of `ID`, feel free to use `ariG23498/intersection-dataset` as the example.

File: count_intersection/README.md [hunk 3]
```diff
@@ -37,22 +39,32 @@ python create_intersection_dataset.py \
...
+    --ckpt_id="google/paligemma2-3b-pt-224" \
+    --push_to_hub=true \
+    --model_id="YOUR_HF_USERNAME/count_intersection-ft-paligemma2-3b-pt-224"
 ```
 
 - **Parameters**:
   - `ckpt-id`: Pre-trained model checkpoint ID.
   - `dataset_folder`: Directory containing the dataset.
+  - `push_to_hub`: Whether to upload the fine-tuned model to the Hugging Face Hub.
+  - `model_id`: ID (username + model name) used to upload to the Hub. Required if `push_to_hub` is True.
```
Comment (ariG23498) [inline count_intersection/README.md]: Same as above. An example of the ID would be helpful for the user.

File: count_intersection/README.md [hunk 4]
```diff
@@ -37,22 +39,32 @@ python create_intersection_dataset.py \
...
 
 ### Example Workflow
-1. Generate the dataset using `create_intersection_dataset.py`.
-2. Fine-tune the model using `finetune.py`.
-3. Run inference to visualize predictions and validate model performance.
+1. You may need to authenticate with `huggingface-cli login` if you haven't already.
```
Comment (ariG23498) [inline count_intersection/README.md:64]: Great addition.

File: count_intersection/finetune.py [hunk 5]
```diff
@@ -69,7 +71,10 @@ def run_inference(val_dataset, processor, model):
     plt.show()
 
 
-def main(ckpt_id: str = "google/paligemma2-3b-pt-224", dataset_folder: str = "dataset"):
+def main(ckpt_id: str = "google/paligemma2-3b-pt-224", dataset_folder: str = "dataset", push_to_hub: bool = False, model_id: str | None = None):
+    if push_to_hub and model_id is None:
+        raise ValueError("model_id cannot be none")
```
Comment (ariG23498) [inline count_intersection/finetune.py]:
```suggestion
        raise ValueError("`model_id` cannot be `None` when you want to push model to Hub")
```

File: count_intersection/create_intersection_dataset.py [hunk 6]
```diff
@@ -58,7 +58,11 @@ def main(
     dataset_folder: str = "dataset",
     dataset_split: str = "train",
     push_to_hub: bool = False,
+    dataset_id: str | None = None,
 ):
+    if push_to_hub and dataset_id is None:
+        raise ValueError("dataset_id cannot be None.")
```
Comment (ariG23498) [inline count_intersection/create_intersection_dataset.py]:
```suggestion
        raise ValueError("`dataset_id` cannot be `None` when you want to push the dataset to Hub")
```

**Review by ariG23498 — CHANGES_REQUESTED** [review]
Thank you for all the work. It is good as is, but left some minor suggestions.

Something you can look into is the `tags` parameter with the `model.push_to_hub`. That would help fine tuned models be tagged with something specific, so that later we can understand how many people used our repo to fine tune their models. WDYT?

Also I think we can keep the notebook as if, it is anyway not meant for direct usage.

**Comment by oliveirabruno01** [comment]
Thank you for your review, @ariG23498. I’ve addressed most of your suggestions, leaving out the `tags` usage and the notebook situation for clarification.

Regarding the notebook, I’m unsure if you meant for it to remain as is or be modified to match the version in my branch.

As for the `tags`, I wanted to clarify with you before proceeding. Would it be acceptable for the `tag` to be the repository name? I also thought it might be useful to include a tag indicating the project task, for example:

```python
if push_to_hub:
    model.push_to_hub(model_id, tags=["fine-tune-paligemma", "count-intersection"])
```

If you could provide some guidance on these points, I’d be happy to continue working on this PR.

Thanks again for your help!

**Comment by ariG23498** [comment]
@oliveirabruno01 great work so far!

> Regarding the notebook, I’m unsure if you meant for it to remain as is or be modified to match the version in my branch.

Let's think this through, I had only created the notebook so that people could see the inference images before and during training. I think we should safely remove the notebook, and introduce a logging package ([`wandb`](https://wandb.ai/site/)) in another PR and start a training run. That way we could share the logs and discuss further about reproducibility and also compare different methods later. WDYT?

> As for the tags

Love the idea of having multiple tags, please go ahead with the suggestion.

**Comment by oliveirabruno01** [comment]
Thank you for the feedback, @ariG23498. Your plan makes sense, and a separate PR for wandb and training runs sounds like a great idea. I appreciate your guidance!

**Comment by ariG23498** [comment]
Hey @oliveirabruno01 do you want me to review the PR?

**Comment by oliveirabruno01** [comment]
Oh, forgot to mention—yep, it’s ready for review now!

**Review by ariG23498 — APPROVED** [review]
LGTM!

**Comment by ariG23498** [comment]
Thank you for your contribution! 🥳

**Pull request merged**